### PR TITLE
Add Source.close/1 callback to prevent hitting file descriptor limits

### PR DESCRIPTION
**Pull request by 1st8**
Hey evadne,

thank you very much for this great package, it works fantastically and is a pleasure to work with.

I found one issue where generating large archives from local files (1000+) would crash with `reason: :emfile` (i.e. too many open files). 
File sources were simply never closed in the Encoder loop.

My fix is to add a new `Source.close/1` callback, which is called after `:eof` is received from the source.
This also seemed to apply to the URL and Random source, so I implemented matching behaviours there, too.

The tests were a little hard to follow for me, so this is still a work in progress and any advice would be appreciated.
I verified that the problem is fixed regarding file sources, but did not test (other than running `mix test` - which is green) anything regarding URL or Random.

Let me know what you think!

Cheers,
Christoph

### CODE REVIEW

**Comment by evadne** [comment]
@1st8 what a good idea

**Comment by evadne** [comment]
@1st8 IMO the way to go is to ensure the `read()` implementation in sources will close whatever internal handle it has

e.g.

```elixir
  @impl Source
  def read(source) do
    case IO.binread(source.device, get_chunk_size()) do
      :eof -> File.close(source.device); :eof
      data -> data
    end
  end
```

since all streams are read once only.

**Comment by evadne** [comment]
I’ve added the logic into #8 and merged

**Comment by 1st8** [comment]
Cool, thanks!
